internal/controller/pdb_to_pdbwatcher_controller.go

```diff
@@ -0,0 +1,179 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// PDBToPDBWatcherReconciler reconciles a PodDisruptionBudget object.
+type PDBToPDBWatcherReconciler struct {
+	client.Client
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=policy,resources=poddisruptionbudgets,verbs=get;list;create;watch
+
+// Reconcile reads the state of the cluster for a PDB and creates/deletes PDBWatchers accordingly.
+func (r *PDBToPDBWatcherReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// Fetch the PodDisruptionBudget object based on the reconcile request
+	var pdb policyv1.PodDisruptionBudget
+	err := r.Get(ctx, req.NamespacedName, &pdb)
+	if err != nil {
+		// If the PDB is deleted, we should delete the corresponding PDBWatcher.
+		// First, check if the PDBWatcher exists
+		var pdbWatcher types.PDBWatcher
+		err := r.Get(ctx, req.NamespacedName, &pdbWatcher)
+		if err == nil {
+			// Delete PDBWatcher if PDB is deleted
+			err := r.Delete(ctx, &pdbWatcher)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("unable to delete PDBWatcher: %v", err)
+			}
+			log.Log.Info("Deleted PDBWatcher", "name", req.NamespacedName)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// If the PDB exists, create a corresponding PDBWatcher if it does not exist
+	var pdbWatcher types.PDBWatcher
+	err = r.Get(ctx, req.NamespacedName, &pdbWatcher)
+
+	if err != nil {
+		deploymentName, e := r.discoverDeployment(ctx, &pdb)
+		if e != nil {
+			return reconcile.Result{}, e
+		}
+
+		// Create a new PDBWatcher
+		pdbWatcher = types.PDBWatcher{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "PDBWatcher",
+				APIVersion: "apps.mydomain.com/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pdb.Name,
+				Namespace: pdb.Namespace,
+				Annotations: map[string]string{
+					"controlledBy": "PDBToPDBWatcherController",
+					"target":       deploymentName,
+				},
+			},
+			Spec: types.PDBWatcherSpec{
+				TargetName: deploymentName,
+				TargetKind: deploymentKind,
+			},
+		}
+
+		err := r.Create(ctx, &pdbWatcher)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("unable to create PDBWatcher: %v", err)
+		}
+		log.Log.Info("Created PDBWatcher", "name", pdb.Name)
+	}
+
+	// Return no error and no requeue
+	return reconcile.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PDBToPDBWatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Set up the controller to watch Deployments and trigger the reconcile function
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&policyv1.PodDisruptionBudget{}).
+		WithEventFilter(predicate.Funcs{
+			// Only trigger for Create and Delete events
+			CreateFunc: func(e event.CreateEvent) bool {
```

Comment:
A reminder thes will be called for every pdbwatcher when we restart Create doesn't mean apiserver create it really mean hey ther's an object that I havne't seen in my controllers lifetime.